lib/src/podman_ostree.rs

```diff
@@ -0,0 +1,296 @@
+
+use anyhow::{Context, Result};
+
+use cap_std::fs::Dir;
+use cap_std::fs::{DirBuilder, DirEntry};
+use cap_std::io_lifetimes::AsFilelike;
+use cap_std_ext::cap_tempfile::{TempDir, TempFile};
+use cap_std_ext::cmdext::CapStdExtCommandExt;
+use cap_std_ext::dirext::CapStdExtDirExt;
+use cap_std_ext::{
+    cap_primitives::{
+        self,
+        fs::{DirBuilderExt, MetadataExt, OpenOptionsExt},
+    },
+    cap_std,
+};
+use fn_error_context::context;
+use ostree_ext::sysroot::SysrootLock;
+use rustix::fd::AsFd;
+
+use crate::deploy::ImageState;
+use crate::podman::PodmanInspectGraphDriver;
+use crate::utils::sync_cmd_in_root;
+
+const OSTREE_CONTAINER_IMAGE_REF_PREFIX: &str = "ostree-container/image";
+
+fn image_commit_ostree_ref(imageid: &str) -> String {
+    format!("{OSTREE_CONTAINER_IMAGE_REF_PREFIX}/{imageid}")
+}
+
+struct MergeState<'a> {
+    trash: &'a Dir,
+    // Unique integer for naming trashed files
+    trashid: AtomicI64,
+    can_clone: bool,
+}
+
+/// Given one directory entry, perform an overlayfs-style merge operation.
+fn merge_one_entry(
+    layer: &Dir,
+    elt: DirEntry,
+    pathbuf: &mut std::path::PathBuf,
+    output: &Dir,
+    state: &MergeState,
+) -> Result<()> {
+    let name = elt.file_name();
+    // We operate on a shared path buffer for improved efficiency.
+    // Here, we append the name of the target file.
+    pathbuf.push(&name);
+    let src_meta = elt.metadata()?;
+    let inum = src_meta.ino();
+    let src_ftype = src_meta.file_type();
+
+    // Helper closure which lazily initializes a "layer trash directory" and moves the target path into it.
+    let move_to_trash = |src: &Path| -> anyhow::Result<()> {
+        let id = state.trashid.fetch_add(1, Ordering::SeqCst);
+        let tempname = format!("t{:X}-{:X}", id, inum);
+        output
+            .rename(src, state.trash, &tempname)
+            .with_context(|| format!("Moving {src:?} to trash"))?;
+        Ok(())
+    };
+
+    let target_meta = output
+        .symlink_metadata_optional(&pathbuf)
+        .context("Querying target")?;
+    if src_ftype.is_dir() {
+        // The source layer type is a directory.  Check if we need to create it.
+        let mut needs_create = true;
+        if let Some(target_meta) = target_meta {
+            if target_meta.is_dir() {
+                needs_create = false;
+            } else {
+                // The target exists and is not a directory.  Trash it.
+                move_to_trash(&pathbuf)?;
+            }
+        }
+        // Create the directory if needed.
+        if needs_create {
+            let mut db = DirBuilder::new();
+            db.mode(src_meta.mode());
+            output
+                .create_dir_with(&pathbuf, &db)
+                .with_context(|| format!("Creating {pathbuf:?}"))?;
+        }
+        // Now recurse
+        merge_layer(layer, pathbuf, output, state)?;
+    } else if (src_meta.mode() & libc::S_IFMT) == libc::S_IFCHR && src_meta.rdev() == 0 {
```

Comment:
are opaque directories handled (or should they)?

Reply:
I think you're right, we need to do that.  Huh.  So I did [this search for Go code referencing trusted.overlay.opaque](https://github.com/search?q=lang%3Ago+trusted.overlay.opaque&type=code), it's pretty interesting all the stuff out there...including [this windows tar2ext4 code](https://github.com/microsoft/hcsshim/blob/8beabacfc2d21767a07c20f8dd5f9f3932dbf305/ext4/tar2ext4/tar2ext4.go).  There's some interesting hits on the Rust side too, like [nydus](https://github.com/dragonflyoss/nydus/blob/master/builder/src/core/tree.rs#L34).